tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -4324,6 +4325,47 @@ void CGMSHLSLRuntime::EmitHLSLDiscard(CodeGenFunction &CGF) {
       TheModule);
 }
 
+// Emit an artificially conditionalized branch for a break operation when in a potentially wave-enabled stage
+// This allows the block containing what would have been an unconditional break to be included in the loop
+// If the block uses values that are wave-sensitive, it needs to stay in the loop to prevent optimizations
+// that might produce incorrect results by ignoring the volatile aspect of wave operation results.
+void CGMSHLSLRuntime::EmitHLSLCondBreak(CodeGenFunction &CGF, Function *F, BasicBlock *DestBB, BasicBlock *AltBB) {
+  // If not a wave-enabled stage, we can keep everything unconditional as before
+  if (!m_pHLModule->GetShaderModel()->IsPS() && !m_pHLModule->GetShaderModel()->IsLib()) {
+    CGF.Builder.CreateBr(DestBB);
+    return;
+  }
+
+  // Create an internal global for the conditional branch to depend on.
+  Constant *GV = TheModule.getGlobalVariable("dx.break");
```

Comment:
Can we place "dx.break" in a constant global variable, instead of defined in place?

Reply:
When I did that, the first simplify-inst pass I hit deconditionalized the breaks. Maybe there's a clever way around that, but I haven't found it.

The reason the dx.preserve change didn't run into this problem was because it was only used on unoptimized runs that never use that pass.

Reply:
The global did get set to const before the end.

With my latest change which the global isn't created at all until near the end when no remaining passes are at risk of optimizing it away.